src/main/java/com/treasuredata/client/TDHttpClient.java

```diff
@@ -303,148 +288,188 @@ public String apply(Integer input)
         }
         for (Map.Entry<String, String> entry : apiRequest.getHeaderParams().entries()) {
-            request.header(entry.getKey(), entry.getValue());
+            request = request.addHeader(entry.getKey(), entry.getValue());
         }
 
         // Submit method specific headers
         switch (apiRequest.getMethod()) {
+            case GET:
+                request = request.get();
+                break;
+            case DELETE:
+                request = request.delete();
+                break;
             case POST:
                 if (apiRequest.getPostJson().isPresent()) {
-                    request.content(new StringContentProvider(apiRequest.getPostJson().get()), "application/json");
+                    request = request.post(RequestBody.create(mediaTypeJson, apiRequest.getPostJson().get()));
                 }
                 else if (queryStr.length() > 0) {
-                    request.content(new StringContentProvider(queryStr), "application/x-www-form-urlencoded");
+                    request = request.post(RequestBody.create(mediaTypeXwwwFormUrlencoded, queryStr));
                 }
                 else {
                     // We should set content-length explicitly for an empty post
-                    request.header("Content-Length", "0");
+                    request = request
+                            .header("Content-Length", "0")
+                            .post(RequestBody.create(null, ""));
                 }
                 break;
             case PUT:
                 if (apiRequest.getPutFile().isPresent()) {
                     try {
-                        request.file(apiRequest.getPutFile().get().toPath(), "application/octet-stream");
+                        request = request.put(RequestBody.create(mediaTypeOctetStream, apiRequest.getPutFile().get()));
                     }
-                    catch (IOException e) {
+                    catch (NullPointerException e) {
                         throw new TDClientException(TDClientException.ErrorType.INVALID_INPUT, "Failed to read input file: " + apiRequest.getPutFile().get());
                     }
                 }
                 break;
         }
 
-        // Configure redirect (302) following
-        Optional<Boolean> followRedirects = apiRequest.getFollowRedirects();
-        if (followRedirects.isPresent()) {
-            request.followRedirects(followRedirects.get());
-        }
+        // OkHttp will follow redirect (302)
 
-        return request;
+        return request.build();
     }
 
     private static boolean isNakedTD1Key(String s)
     {
         return NAKED_TD1_KEY_PATTERN.matcher(s).matches();
     }
 
-    public <ResponseType extends Response, Result> Result submitRequest(TDApiRequest apiRequest, Optional<String> apiKeyCache, Handler<ResponseType, Result> handler)
-            throws TDClientException
+    private class RequestContext
```

Comment:
TODO: This should be a static class. We only need to pass some config parameters when initializing this class.